samples/bluetooth/periodic_adv_rsp/src/main.c

```diff
@@ -0,0 +1,388 @@
+ */
+
+#include <zephyr/bluetooth/att.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+
+#define NUM_RSP_SLOTS 5
+#define NUM_SUBEVENTS 5
+#define PACKET_SIZE   5
+#define NAME_LEN      30
+
+static K_SEM_DEFINE(sem_connected, 0, 1);
+static K_SEM_DEFINE(sem_discovered, 0, 1);
+static K_SEM_DEFINE(sem_written, 0, 1);
+static K_SEM_DEFINE(sem_disconnected, 0, 1);
+
+static struct bt_uuid_128 pawr_char_uuid =
+	BT_UUID_INIT_128(BT_UUID_128_ENCODE(0x12345678, 0x1234, 0x5678, 0x1234, 0x56789abcdef1));
+static uint16_t pawr_attr_handle;
+static const struct bt_le_per_adv_param per_adv_params = {
+	.interval_min = 0xFF,
+	.interval_max = 0xFF,
+	.options = 0,
+	.num_subevents = NUM_SUBEVENTS,
+	.subevent_interval = 0x30,
+	.response_slot_delay = 0x5,
+	.response_slot_spacing = 0x50,
+	.num_response_slots = NUM_RSP_SLOTS,
+};
+
+static struct bt_le_per_adv_subevent_data_params subevent_data_params[NUM_SUBEVENTS];
+static struct net_buf_simple bufs[NUM_SUBEVENTS];
+static uint8_t backing_store[PACKET_SIZE][NUM_SUBEVENTS];
+
+BUILD_ASSERT(ARRAY_SIZE(bufs) == ARRAY_SIZE(subevent_data_params));
+BUILD_ASSERT(ARRAY_SIZE(backing_store) == ARRAY_SIZE(subevent_data_params));
+
+static uint8_t counter;
+
+static void request_cb(struct bt_le_ext_adv *adv, const struct bt_le_per_adv_data_request *request)
+{
+	int err;
+	uint8_t to_send;
+	struct net_buf_simple *buf;
+
+	to_send = MIN(request->count, ARRAY_SIZE(subevent_data_params));
+
+	for (size_t i = 0; i < to_send; i++) {
+		buf = &bufs[i];
+		buf->data[buf->len - 1] = counter++;
+
+		subevent_data_params[i].subevent =
+			(request->start + i) % per_adv_params.num_subevents;
+		subevent_data_params[i].response_slot_start = 0;
+		subevent_data_params[i].response_slot_count = NUM_RSP_SLOTS;
+		subevent_data_params[i].data = buf;
+	}
+
+	err = bt_le_per_adv_set_subevent_data(adv, to_send, subevent_data_params);
+	if (err) {
+		printk("Failed to set subevent data (err %d)\n", err);
+	}
+}
+
+static bool print_ad_field(struct bt_data *data, void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	printk("    0x%02X: ", data->type);
+	for (size_t i = 0; i < data->data_len; i++) {
+		printk("%02X", data->data[i]);
+	}
+
+	printk("\n");
+
+	return true;
+}
+
+static struct bt_conn *default_conn;
+
+static void response_cb(struct bt_le_ext_adv *adv, struct bt_le_per_adv_response_info *info,
+		     struct net_buf_simple *buf)
+{
+	if (buf) {
+		printk("Response: subevent %d, slot %d\n", info->subevent, info->response_slot);
+		bt_data_parse(buf, print_ad_field, NULL);
+	} else {
+		printk("Failed to receive response: subevent %d, slot %d\n", info->subevent,
+		       info->response_slot);
+	}
+}
+
+static const struct bt_le_ext_adv_cb adv_cb = {
+	.pawr_data_request = request_cb,
+	.pawr_response = response_cb,
+};
+
+void connected(struct bt_conn *conn, uint8_t err)
+{
+	printk("Connected (err 0x%02X)\n", err);
+
+	if (err) {
+		bt_conn_unref(conn);
```

Comment:
This doesn't look correct. You don't own the `conn` reference in the connected callback, so you're not allowed to unref it either. If you want to keep a reference to it after the callback returns, then you need to take your own reference using `bt_conn_ref()`

Reply:
I agree. @hermabe The best approach is to tie the ownership of a reference to a variable. I just now made some wrapping functions around `bt_conn_ref`/`bt_conn_unref` that I think may be a safer interface.

```c
void conn_ref_store(struct bt_conn **conn_var, struct bt_conn *owned)
{
	void *conn_var_before = atomic_ptr_set((void **)conn_var, owned);
	__ASSERT_NO_MSG(conn_var_before == NULL);
}

struct bt_conn *__must_check conn_ref_move(struct bt_conn **conn_var)
{
	return atomic_ptr_clear((void **)conn_var);
}

void conn_ref_clone(struct bt_conn **conn_var, struct bt_conn *borrowed)
{
	struct bt_conn *owned = bt_conn_ref(borrowed);
	conn_ref_store(conn_var, owned);
}

void conn_ref_clear(struct bt_conn **conn_var)
{
	struct bt_conn *owned = conn_ref_move(conn_var);
	bt_conn_unref(owned);
}
```

Normal usage is trough `conn_ref_clone` and `conn_ref_clear`. `conn_ref_store` would only be used directly in special cases when you obtained an owned reference in a return value from somewhere, since you cannot take the address of a return value.

Reply:
I have updated it to be more explicit. `bt_conn_le_create` gives a reference, which is unrefed in the failure cases. `conn` can never be `NULL` in the connected callback as all paths calling it dereference `conn`.